network/cloudflare_dns.py

```diff
@@ -0,0 +1,448 @@
+
+def main():
+    module = AnsibleModule(
+        argument_spec = dict(
+            account_api_token = dict(required=True, default=None, no_log=True, type='str'),
+            account_email     = dict(required=True, default=None, type='str'),
+            record            = dict(required=False, default=None, aliases=['name'], type='str'),
+            timeout           = dict(required=False, default=30, type='int'),
+            ttl               = dict(required=False, default=1, type='int'),
+            type              = dict(required=False, default=None, choices=[ 'A', 'AAAA', 'CNAME', 'TXT', 'SRV', 'LOC', 'MX', 'NS', 'SPF' ], type='str'),
+            solo              = dict(required=False, default=None, type='bool'),
+            state             = dict(required=False, default='present', choices=['present', 'absent'], type='str'),
+            value             = dict(required=False, default=None, aliases=['content'], type='str'),
+            zone              = dict(required=False, default=None, aliases=['domain'], type='str'),
+        ),
+        supports_check_mode = True,
+    )
+
+    changed = False
+    cf_api = CloudflareAPI(module)
+
+    # sanity checks
+    if cf_api.record == '*' and cf_api.state == 'present':
+        module.fail_json(msg="record=* can only be used with state=absent")
+    if cf_api.is_solo and cf_api.state == 'absent':
+        module.fail_json(msg="solo=true can only be used with state=present")
+
+    # No zone specified
+    # -> return zone information for available zones
+    if (not cf_api.zone):
+        zones = cf_api.get_zones()
+        module.exit_json(changed=changed, result={'zones': zones})
+
+    # Zone is specified, but not record, value and type
+    # -> return all records for that zone
+    if cf_api.zone and not cf_api.record and not cf_api.value and not cf_api.type:
+        records = cf_api.get_dns_records()
+        module.exit_json(changed=changed, result={'records': records})
+
+    # Zone is specified and at least one of record, value or type has been provided
+    # -> perform add, delete or update (only the TTL can be updated)
+    if cf_api.zone and ((cf_api.record is not None) or (cf_api.value is not None) or (cf_api.type is not None)):
+        if cf_api.state == 'present':
+            # delete all records matching record name + type
+            if cf_api.is_solo:
+                changed = cf_api.delete_dns_records(solo=cf_api.is_solo)
+            result,changed = cf_api.create_dns_record()
+            module.exit_json(changed=changed,result={'records': result})
+        elif cf_api.state == 'absent':
+            # force solo to False, just to be sure
+            changed = cf_api.delete_dns_records(solo=False)
+            module.exit_json(changed=changed)
+        else:
+            module.fail_json(msg="{0} is an unknown state".format(cf_api.state))
+
+    module.fail_json(msg="Unexpected combination of zone, record, value and type.")
```

Comment:
I was initially considering these but did not use them as the parameters are only required together or mutually exclusive under certain conditions. The mutually_exclusive and required_together parameters have to be provided during the instantiation of the AnsibleModule class. I would need something "require_oneof_ if" or "conflicts_if". Please give me  a pointer to where I have to look if I have overlooked something.